src/main/java/bowling/Frame.java

```diff
@@ -0,0 +1,119 @@
+package bowling;
+
+import bowling.exception.BowlingException;
+import bowling.exception.FrameOverException;
+import bowling.exception.InvalidPinQuantityException;
+
+public class Frame {
+    public static final int MAX_PINS = 10;
+    private final boolean isFinalFrame;
+    private int pinsStanding = MAX_PINS;
+    private int rollNumber = 0;
+    private boolean isOver = false;
+    private int rollOnePins = 0;
+    private int rollTwoPins = 0;
+    private FrameType frameType = FrameType.NORMAL;
+    private int score = 0;
+
+    public Frame(boolean isFinalFrame) {
+        this.isFinalFrame = isFinalFrame;
+    }
+
+    public Frame() {
+        this(false);
+    }
+
+    public void roll(int numberOfPinsKnockedDown) throws BowlingException {
+        if (isOver) {
+            throw new FrameOverException("This frame is over, wyd fam");
+        }
+
+        if (numberOfPinsKnockedDown > pinsStanding) {
+            throw new InvalidPinQuantityException("R u dizzy u knocked over  more pins than there r");
+        }
+
+        rollNumber++;
+        pinsStanding -= numberOfPinsKnockedDown;
+
+        if (isFinalFrame) {
+            rollFinalFrame(numberOfPinsKnockedDown);
+        } else {
+            rollNonFinalFrame(numberOfPinsKnockedDown);
+        }
+    }
+
+    public void rollNonFinalFrame(int numberOfPinsKnockedDown) {
```

Comment:
These should not be public since they are invoked by the `roll` method.

Reply:
Also, as a personal preference, I am not a huge fan of modifying a lot of state in methods. Conceptually I get it. The class is reflecting a game and the methods are domain intuitive. But personally I like to see at a quick glance of a function how it is changing state. So in this case I noticed it depends on `rollNumber` and I thought it might make sense to have that passed as an argument, but actually it is modifying `frameType` and `isOver` as well. Perhaps those properties could be returned as values from the function? Then you could have a `private static` or a separate class that is responsible for these state calculations. Making them easier to test in isolation.